.adr/0017-missed-chain-events.md

```diff
@@ -0,0 +1,127 @@
+
+## Decision
+
+After considering the aforementioned alternatives, the following design decisions were made.
+
+### Avoid init race conditions that would cause missed events
+
+To ensure no events are missed or processed out of order during the `chainservice` initialization, we can modify the `chainservice` init routine to acquire/release the `eventTracker`'s `mutex` as shown below:
+
+```go
+	// Prevent go routines from processing events before checkForMissedEvents completes
+	ecs.eventTracker.mu.Lock()
+	{
+		ecs.wg.Add(3)
+		go ecs.listenForEventLogs(errChan, eventSub, eventChan, eventQuery)
+		go ecs.listenForNewBlocks(errChan, newBlockSub, newBlockChan)
+		go ecs.listenForErrors(errChan)
+
+		// Search for any missed events emitted while this node was offline
+		err = ecs.checkForMissedEvents(startBlock)
+		if err != nil {
+			return nil, err
+		}
+	}
+	ecs.eventTracker.mu.Unlock()
+```
+
+The `listenForEventLogs` go routine also tries to acquire the `mutex` before updating the `eventTracker.events` queue but will be blocked until `checkForMissedEvents` completes. This means that `listenForEventLogs` will still be able to detect new chain events, but will not add them to the queue or trigger event processing until `checkForMissedEvents` finishes queueing old events.
+
+### 1. Put `Channel` data protection logic in `Channel` class
+
+Add logic to the `Channel` class to protect its data from stale updates rather than relying on an external component (`chainservice`) to only feed it sanitized data. This seems less error-prone and it could be useful to add a `Channel.LastUpdated` struct in case a user wants to know when the channel was last changed. The `chainservice` will still be expected to feed chain events in order, but the added `Channel` logic will act as another layer of protection against faulty data updates.
+
+The `Channel` class should keep track of the `TxIndex` associated with each chain event in case there are multiple events related to a single channel within the same block. Also, an `ethereum.FilterQuery` cannot specify a `TxIndex` so it will be more computationally efficient to check the `TxIndex` in the `Channel` class rather than in the `chainservice` logic.
+
+```go
+type Channel struct {
+	state.FixedPart
+	Id      types.Destination
+	MyIndex uint
+
+	OnChain  OnChainData
+	OffChain OffChainData
+
+  LastChainUpdate ChainUpdateData
+}
+
+type ChainUpdateData struct {
+        BlockNum uint64
+        TxIndex uint64
+}
+
+func (c *Channel) UpdateWithChainEvent(event chainservice.Event) (*Channel, error) {
+	if event.BlockNum() > c.LastChainUpdate.BlockNum && event.TxIndex() > c.LastChainUpdate.TxIndex {
```

Comment:
I think this check might have to be tweaked a bit.

- `event.BlockNum() > c.LastChainUpdate.BlockNum` we're ignoring the case where `event.BlockNum==c.LastChainUpdate.BlockNum` which is where `TxIndex` comes into play
- We require the `TxIndex` always be larger then the last, even if the events are in different blocks.
I think we want something like:

```suggestion
	if event.BlockNum() > c.LastChainUpdate.BlockNum || (event.BlockNum() == c.LastChainUpdate.BlockNum) && event.TxIndex() > c.LastChainUpdate.TxIndex {
```

Reply:
Great catch! I will work on a PR that implements the `Channel` class data protections mentioned above and link that here before merging this adr PR

Reply:
PR to make the above modifications to the `Channel` class: https://github.com/statechannels/go-nitro/pull/1659